File: E4.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <stack>
#include <map>
#include <algorithm>
#include <iomanip>

using namespace std;

// Função DFS para ordenação topológica
void dfs(int idx, const vector<vector<int>>& adj, vector<bool>& visitado, stack<int>& s) {
    visitado[idx] = true;
    for (int idxNew = 0; idxNew < adj[idx].size(); ++idxNew) {
        if (adj[idx][idxNew] && !visitado[idxNew]) {
            dfs(idxNew, adj, visitado, s);
        }
    }
    s.push(idx);
}

vector<int> ordTopologica(const vector<vector<int>>& adj){
    int n = adj.size();
    vector<bool> visitado(n, false);
    stack<int> s;

    for (int i = 0; i < n; ++i) {
        if (!visitado[i]) dfs(i, adj, visitado, s);
    }

    vector<int> ordem;
    while (!s.empty()) {
        ordem.push_back(s.top());
        s.pop();
    }
    return ordem;
}

int main(){
    int nTestes;
    cin >> nTestes;

    // Tabela de produtividade dos heróis
    map<string, double> niveisProd = {{"Lenda", 2.0}, {"Mestre", 1.5}, {"Cavaleiro", 1.2}, {"Aventureiro", 1.0}, {"Aprendiz", 0.75}};

    for(int i = 0; i < nTestes; i++){
        int nHerois, nQuests;
        cin >> nHerois >> nQuests;

        // Leitura dos heróis:
        vector<pair<string, string>> listaHerois;   // (nome, nivel)
        for(int j = 0; j < nHerois; j++){
            string heroi, nivel;
            cin >> heroi >> nivel;

            listaHerois.push_back({heroi, nivel});
        }

        cin.ignore();   // limpar buffer antes de getline

        // Leitura das Quests e dependências
        vector<pair<int, int>> listaQuests;                                 // (id, tempoExec)
        vector<vector<int>> matrizDpQuests(nQuests, vector<int>(nQuests, 0));
        for(int j = 0; j < nQuests; j++){
            string linhaQuest;
            getline(cin, linhaQuest);
            istringstream ss(linhaQuest);

            int id, tempoExec, dp;
            ss >> id >> tempoExec;
            listaQuests.push_back({id, tempoExec});

            while(ss >> dp){
                // marca 1 na matriz de dependencias
                if(dp != 0){
                    matrizDpQuests[id-1][dp-1] = 1;
                }
            }
        }

        // Ordenação Topológica das quests
        vector<int> ordemQuests = ordTopologica(matrizDpQuests);

        // Inicializa o tempo livre de cada heroi
        vector<double> tempoLivreHeroi(nHerois, 0);
        vector<double> tempoConcluidoQuest(nQuests, 0);
        vector<vector<int>> questsPorHeroi(nHerois); // quais quests cada herói fez

        // Algoritmo guloso
        for(int q : ordemQuests){
            int tempoBase = listaQuests[q].second;

            // calcular tempo mínimo devido às dependências
            double tempoMinimoParaQuest = 0;
            for(int i = 0; i < nQuests; i++){
                if(matrizDpQuests[i][q]) // i antes de q
                    tempoMinimoParaQuest = max(tempoMinimoParaQuest, tempoConcluidoQuest[i]);
            }

            double menorFim = 1e9;
            int heroiEscolhido = -1;

            for(int h = 0; h < nHerois; h++){
                double tempoInicio = max(tempoLivreHeroi[h], tempoMinimoParaQuest);
                double tempoFim = tempoInicio + tempoBase / niveisProd[listaHerois[h].second];
                if(tempoFim < menorFim){
                    menorFim = tempoFim;
                    heroiEscolhido = h;
                }
            }

            // Atualiza
            tempoLivreHeroi[heroiEscolhido] = menorFim;
            tempoConcluidoQuest[q] = menorFim;
            questsPorHeroi[heroiEscolhido].push_back(q+1); // armazenar ID da quest
        }

        // Saída
        cout << fixed << setprecision(2);
        for(int h = 0; h < nHerois; h++){
            sort(questsPorHeroi[h].begin(), questsPorHeroi[h].end());
            cout << listaHerois[h].first << " = {";
            for(size_t k = 0; k < questsPorHeroi[h].size(); k++){
                cout << questsPorHeroi[h][k];
                if(k+1 < questsPorHeroi[h].size()) cout << ",";
            }
            cout << "}\n";
        }

        double tempoTotal = *max_element(tempoLivreHeroi.begin(), tempoLivreHeroi.end());
        cout << "Tempo mínimo: " << tempoTotal << endl;

    }

    return 0;    
}

File: E4.py
from collections import defaultdict, deque

# Produtividade dos heróis
produtividade = {
    "Aprendiz": 0.75,
    "Aventureiro": 1.0,
    "Cavaleiro": 1.2,
    "Mestre": 1.5,
    "Lenda": 2.0
}

nTestes = int(input().strip())
for i in range(nTestes):
    nHerois, nQuests = map(int, input().split())
    
    # Ler heróis
    herois = []
    for _ in range(nHerois):
        nome, nivel = input().split()
        herois.append({
            "nome": nome,
            "nivel": nivel,
            "prod": produtividade[nivel],
            "tempo_livre": 0.0,
            "quests": []
        })
    
    # Ler quests
    tempo_base = {}
    dependencias_q = defaultdict(list)
    indegree = [0] * (nQuests+1)
    
    for _ in range(nQuests):
        linha = list(map(int, input().split()))
        id, tempo, *deps = linha
        tempo_base[id] = tempo
        if deps[0] != 0:  # se há dependências
            dependencias_q[id] = deps
            for d in deps:
                indegree[id] += 1
        else:
            dependencias_q[id] = []
    
    # Ordenação topológica (Kahn)
    fila = deque([i for i in range(1, nQuests+1) if indegree[i] == 0])
    ordem = []
    while fila:
        atual = fila.popleft()
        ordem.append(atual)
        for q, deps in dependencias_q.items():
            if atual in deps:
                indegree[q] -= 1
                if indegree[q] == 0:
                    fila.append(q)
    
    # Simulação
    fim_quest = {}
    for q in ordem:
        melhor_fim = float("inf")
        escolhido = None
        for h in herois:
            inicio = max(h["tempo_livre"], max((fim_quest[d] for d in dependencias_q[q]), default=0))
            duracao = tempo_base[q] / h["prod"]
            fim = inicio + duracao
            if fim < melhor_fim:
                melhor_fim = fim
                escolhido = h
        # Atribui quest ao herói
        escolhido["tempo_livre"] = melhor_fim
        escolhido["quests"].append(q)
        fim_quest[q] = melhor_fim
    
    # Saída
    for h in herois:
        quests_str = ",".join(map(str, h["quests"]))
        print(f"{h['nome']} = {{{quests_str}}}")
    print(f"Tempo mínimo: {max(fim_quest.values()):.2f}")
    if i != nTestes-1:
        print()